### PR TITLE
[Graphbolt]Add negative sampler udf

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
UDF of negative sampler in graphbolt.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/negative_sampler.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        self.negative_ratio = negative_ratio
+
+    def _sample(self, node_pairs):
+        """
+        Generates a mix of positive and negative samples.
+
+        Parameters
+        ----------
+        node_pairs : Tuple[Tensor] or Dict[etype, Tuple[Tensor]]
+            A tuple of tensors or a dictionary that represents source-destination node pairs
```
Comment (frozenbugs) [inline python/dgl/graphbolt/negative_sampler.py:45]: line too long.

File: python/dgl/graphbolt/impl/uniform_negative_sampler.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        """
+        Initlization for a uniform negative sampler.
+
+        Parameters
+        ----------
+        datapipe : DataPipe
+            The datapipe.
+        negative_ratio : int
+            The proportion of negative samples to positive samples.
+        graph
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/uniform_negative_sampler.py:31]: Worth yo call out the supported Graph format. For now, just CSCSamplingGraph.

File: python/dgl/graphbolt/impl/uniform_negative_sampler.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            The graph on which to perform negative sampling.
+        """
+        super().__init__(datapipe, negative_ratio)
+        self.graph = graph
+
+    def _generate_negative_pairs(self, node_pairs, etype=None):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/uniform_negative_sampler.py:37]:
```suggestion
    def _sample_negative_pairs(self, node_pairs, etype=None):
```

File: python/dgl/graphbolt/negative_sampler.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        etype : (str, str, str)
+            Canonical edge type.
+        Returns
+        -------
+        Tuple[Tensor]
+            A mixed collection of positive and negative node pairs.
+        """
+        raise NotImplementedError
+
+    def _generate_negative_pairs(self, node_pairs, etype=None):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/negative_sampler.py:77]:
```suggestion
    def _sample_negative_pairs(self, node_pairs, etype=None):
```

File: python/dgl/graphbolt/negative_sampler.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        """
+        if isinstance(node_pairs, Mapping):
+            return {
+                etype: self._generate(pos_pairs, etype)
+                for etype, pos_pairs in node_pairs.items()
+            }
+        else:
+            return self._generate(node_pairs, None)
+
+    def _generate(self, node_pairs, etype=None):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/negative_sampler.py:59]:
```suggestion
    def _sample_with_etype(self, node_pairs, etype=None):
```

File: python/dgl/graphbolt/negative_sampler.py (hunk 4 above)
Comment (frozenbugs) [inline python/dgl/graphbolt/negative_sampler.py:77]: def _post_process():
    """Covert negative pairs to a specified format.

File: python/dgl/graphbolt/negative_sampler.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+class IndependentNegativeSampler(NegativeSampler):
+    """
+    A kind of negative sampler. `Independent` denotes the data format, where
+    data is structured as triples `[u, v, label]`. Each triple represents an
+    edge between nodes 'u' and 'v', and the 'label' indicates whether the edge
+    is negative (0) or positive (1).
+    """
+
+    def _generate(self, node_pairs, etype=None):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/negative_sampler.py:103]: def _post_process:

File: python/dgl/graphbolt/negative_sampler.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            A collection of edges or a dictionary that maps etypes to edges,
+            which includes both positive and negative samples.
+        """
+        if isinstance(node_pairs, Mapping):
+            return {
+                etype: self._generate(pos_pairs, etype)
+                for etype, pos_pairs in node_pairs.items()
+            }
+        else:
+            return self._generate(node_pairs, None)
```
Comment (frozenbugs) [inline python/dgl/graphbolt/negative_sampler.py:57]: return self._post_process(self._generate(node_pairs, None))

File: python/dgl/graphbolt/negative_sampler.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            Canonical edge type.
+        Returns
+        -------
+        Tuple[Tensor]
+            A collection of negative node pairs.
+        """
+        raise NotImplementedError
+
+
+class IndependentNegativeSampler(NegativeSampler):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/negative_sampler.py:95]: Is independent a good name? Can you confirm with @BarclayII , same to Conditioned.

**Comment by frozenbugs** [comment]
Please also add unit test.

File: python/dgl/graphbolt/negative_sampler.py (hunk 4 above)
Comment (peizhou001) [reply python/dgl/graphbolt/negative_sampler.py:77]: It has conflict with above comment which suggest using name `_sample_negative_pairs`. If so, I prefer `_post_process` as it is more accurate.

File: python/dgl/graphbolt/negative_sampler.py (hunk 8 above)
Comment (peizhou001) [reply python/dgl/graphbolt/negative_sampler.py:95]: In previous doc, these name are suggested.

File: python/dgl/graphbolt/negative_sampler.py (hunk 5 above)
Comment (peizhou001) [reply python/dgl/graphbolt/negative_sampler.py:59]: Not very clear about it, the below comment suggest using `_post_process`.

File: python/dgl/graphbolt/negative_sampler.py [hunk 9]
```diff
@@ -50,32 +50,40 @@ def _sample(self, node_pairs):
         """
         if isinstance(node_pairs, Mapping):
             return {
-                etype: self._generate(pos_pairs, etype)
+                etype: self._collate(
+                    pos_pairs, self._sample_negative_pairs(pos_pairs, etype)
```
Comment (frozenbugs) [inline python/dgl/graphbolt/negative_sampler.py:54]: sample_with_etype is probably better.

File: python/dgl/graphbolt/negative_sampler.py [hunk 10]
```diff
@@ -50,32 +50,40 @@ def _sample(self, node_pairs):
...
             }
         else:
-            return self._generate(node_pairs, None)
+            return self._collate(
+                node_pairs, self._sample_negative_pairs(node_pairs, None)
+            )
 
-    def _generate(self, node_pairs, etype=None):
-        """Generate a mix of positive and negative samples for a given etype.
+    def _collate(self, pos_pairs, neg_pairs):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/negative_sampler.py:63]: nit: swap the order of _collate and _sample_negative_pairs.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/negative_sampler.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+            Canonical edge type.
+
+        Returns
+        -------
+        Tuple[Tensor]
+            A collection of negative node pairs.
+        """
+
+
+class IndependentNegativeSampler(NegativeSampler):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/negative_sampler.py:103]: If we only have 2 types of output, I feel the code will be a lot simpler if we merge them and input as parameter.

Comment (peizhou001) [reply python/dgl/graphbolt/negative_sampler.py:103]: make sense, changed for fix

File: python/dgl/graphbolt/link_data_format.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
+"""Linked data format."""
```
Comment (frozenbugs) [inline python/dgl/graphbolt/link_data_format.py:1]: Rename the file to data_format.py
```suggestion
"""Data format enums for graphbolt."""
```

File: python/dgl/graphbolt/link_data_format.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+from enum import Enum
+
+__all__ = ["LinkDataFormat"]
+
+
+class LinkDataFormat(Enum):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/link_data_format.py:8]: LinkPredictionEdgeFormat

File: python/dgl/graphbolt/link_data_format.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+    """
+    An Enum class representing the two data formats used in link prediction:
```
Comment (frozenbugs) [inline python/dgl/graphbolt/link_data_format.py:10]: If we are 100% sure there are 2 data formats, then we should use boolean, if we use enum, which means there are chances to extend this, then don't mention 2 data formats.
```suggestion
    An Enum class representing the data formats used in link prediction:
```

**Pull request merged**